Add the ability to read from Pandora DB

gwwatkin opened the pull request:
Reads from the padora DB: https://github.com/ioanamoflic/pandora

Uses the flag `--pandora` to specify the localhost port of the Postgres DB instance

mitjarislakki commented on include/lsqecc/patches/patches.hpp:
```diff
@@ -129,9 +133,11 @@ struct Patch {
+
     // Patch& operator=(const Patch& other)
 
+    std::optional<OpId> routing_region_id = -1;// wtf
```
Shouldn't this be an uninitialized optional? This seems to only be used to check whether it was given another value..

mitjarislakki commented on include/pqpp/pq.hpp:
Using this uncommon single header library might lead to maintenance burden down the line compared to something more standard like https://github.com/jtv/libpqxx

& the MIT license should probably be a code comment at the top of the file to ensure compliance
> The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

mitjarislakki committed:
Change -q to general input format argument -I

Change -q to a mutually exclusive general input format with a respective enum for matching
to help readability and adding more formats in the future. Update README to reflect
changes

BREAKING CHANGE: previous `-q` or `--qasm` must now be `--input-format q`,
`--input-format qasm`, `-I q` or `-I qasm`

Signed-off-by: Mitja Rislakki <57709659+mitjarislakki@users.noreply.github.com>

mitjarislakki committed:
Fix wrong indentation

Signed-off-by: Mitja Rislakki <57709659+mitjarislakki@users.noreply.github.com>

mitjarislakki committed:
Add the ability to read from Pandora DB

Rebased & tidied up 8cdbc7820cbd369c966709ecba0e8e73f2fa7f13

Adds new Pandora input format for reading from Postgres DB with
respective GateStream class utilizing gen-op-ids flag.

Signed-off-by: Mitja Rislakki <57709659+mitjarislakki@users.noreply.github.com>

mitjarislakki committed:
feat: add the ability to read from Pandora DB

Rebased & tidied up 8cdbc7820cbd369c966709ecba0e8e73f2fa7f13

Adds new Pandora input format for reading from Postgres DB with
respective GateStream class utilizing gen-op-ids flag.

Skip pq for WASM build

Signed-off-by: Mitja Rislakki <57709659+mitjarislakki@users.noreply.github.com>

mitjarislakki committed:
Sync readme usage with source & spec

README and --help had mismatches stemming from previous PRs
changing one but not the other.

Signed-off-by: Mitja Rislakki <57709659+mitjarislakki@users.noreply.github.com>

mitjarislakki committed:
Update regression tests to reflect change from -q to -I qasm

Signed-off-by: Mitja Rislakki <57709659+mitjarislakki@users.noreply.github.com>

mitjarislakki committed:
Update regression tests to reflect change from -q to -I qasm

Signed-off-by: Mitja Rislakki <57709659+mitjarislakki@users.noreply.github.com>

mitjarislakki committed:
fix(pandora): match gate translator

reversed CNOT control and target ordering
should be fixed for whole code in #144

Signed-off-by: Mitja Rislakki <57709659+mitjarislakki@users.noreply.github.com>

mitjarislakki committed:
fix(pandora): match gate translator

Signed-off-by: Mitja Rislakki <57709659+mitjarislakki@users.noreply.github.com>

mitjarislakki commented on src/gates/postgres_gate_stream.cpp:
```diff
@@ -0,0 +1,125 @@
+                    } else if (angle == -0.5) {
+                        return gates::SDg(row["target_q"].get<PostgresBigIntApprox>());
+                    } else if (angle == 0.25) {
+                        return gates::T(row["target_q"].get<PostgresBigIntApprox>());
```
Will create follow-up

mitjarislakki commented on src/gates/postgres_gate_stream.cpp:
```diff
@@ -0,0 +1,125 @@
+    , qreg_(Qreg{"pandora_reg", [&]() -> QubitNum {
+        auto r = db_->exec("select count(*) from layered_cliff_t where layered_cliff_t.type=0;");
+        auto count = r.at(0).at("count").get<int64_t>();
+        if(count > std::numeric_limits<QubitNum>::max())
```
Will create follow-up

mitjarislakki commented on src/gates/postgres_gate_stream.cpp:
```diff
@@ -0,0 +1,125 @@
+    for (auto& row : r) {
+        auto gate =[&]() -> std::optional<gates::Gate> {
+            switch(static_cast<PandoraGateTranslator>(row["type"].get<int>())) {
+                // Only handle supported gates:
+                // TODO here we are pretending X and Y gates are actually Z, for simplicity. This is not correct, but in 
+                // The future, Pandora will be configured to only output gates supported by lsqecc.
+                case PandoraGateTranslator::XPowGate:
+                case PandoraGateTranslator::YPowGate:
+                case PandoraGateTranslator::ZPowGate:
+                {
+                    double angle = row["param"].get<double>();
+                    if (std::abs(angle) == 1) {
+                        return gates::Z(row["target_q"].get<PostgresBigIntApprox>());
+                    } else if (angle == 0.5) {
+                        return gates::S(row["target_q"].get<PostgresBigIntApprox>());
+                    } else if (angle == -0.5) {
+                        return gates::SDg(row["target_q"].get<PostgresBigIntApprox>());
+                    } else if (angle == 0.25) {
+                        return gates::T(row["target_q"].get<PostgresBigIntApprox>());
+                    } else if (angle == -0.25) {
+                        return gates::TDg(row["target_q"].get<PostgresBigIntApprox>());
+                    } else {
+                        throw std::runtime_error(lstk::cat("Unsupported angle for ZPowGate: ", angle));
+                    }
+                }
+                case PandoraGateTranslator::HPowGate:
+                    return gates::H(row["target_q"].get<PostgresBigIntApprox>());
+                case PandoraGateTranslator::_PauliX:
+                    return gates::X(row["target_q"].get<PostgresBigIntApprox>());
+                case PandoraGateTranslator::_PauliZ:
+                    return gates::Z(row["target_q"].get<PostgresBigIntApprox>());
+                case PandoraGateTranslator::Out:
+                case PandoraGateTranslator::In:
+                case PandoraGateTranslator::M: // TODO implement
+                case PandoraGateTranslator::ResetChannel: // TODO implement
+                    return std::nullopt;
+                case PandoraGateTranslator::CZPowGate:  // TODO: pretending this is a CNOT
+                case PandoraGateTranslator::CXPowGate: // Assume the exponent is 1
```
Will create follow-up

mitjarislakki commented on src/gates/postgres_gate_stream.cpp:
```diff
@@ -0,0 +1,125 @@
+                    return std::nullopt;
+                case PandoraGateTranslator::CZPowGate:  // TODO: pretending this is a CNOT
+                case PandoraGateTranslator::CXPowGate: // Assume the exponent is 1
+                    return gates::CNOT(row["control_q"].get<PostgresBigIntApprox>(),row["target_q"].get<PostgresBigIntApprox>());
```
#144
